allennlp/tests/end_to_end/empty_list_test.py

```diff
@@ -0,0 +1,184 @@
+        embedded_body_tokens = self._embedder(body)
+        # size: <batch_size, 2 * emb_dim>
+        embedded_body = self.fixed_length_embedding(body_mask, embedded_body_tokens)
+
+        # Embed name (in isolation)
+
+        # size: <batch_size, sequence_len>
+        name_mask = get_text_field_mask(entity_name)
+        # size: <batch_size, sequence_len, emb_dim>
+        embedded_name_tokens = self._embedder(entity_name)
+        # size: <batch_size, 2 * emb_dim>
+        embedded_name = self.fixed_length_embedding(name_mask, embedded_name_tokens)
+
+        # Extract embedded spans from the body
+
+        extractor = EndpointSpanExtractor(input_dim=embedded_body_tokens.size(-1))
+        # size: <batch_size, mentions_count>
+        span_mask = (entity_spans[:, :, 0] >= 0).long()
+        # size: <batch_size, mentions_count, 2 * emb_dim>
+        embedded_spans = extractor(embedded_body_tokens, entity_spans, span_indices_mask=span_mask)
+
+        # size: <batch_size>
+        name_match_score = torch.nn.functional.cosine_similarity(embedded_body, embedded_name)
+
+        # size: <batch_size, 2 * emb_dim, mentions_count>
+        transposed_embedded_spans = embedded_spans.transpose(1, 2)
+        # Note: Real model normalizes to give cosine similarity.
+        # size: <batch_size, mentions_count>
+        span_match_scores = torch.bmm(embedded_body.unsqueeze(1), transposed_embedded_spans).squeeze(1)
+        # size: <batch_size, mentions_count>
+        masked_span_match_scores = span_match_scores * span_mask.float()
+        # Aggregate with max to get single score
+        # size: <batch_size>
+        span_match_score = masked_span_match_scores.max(dim=-1)[0].squeeze(-1)
+
+        # Combine name match and span match scores.
+        return {'score': name_match_score + span_match_score}
+
+class EmptyListTest(AllenNlpTestCase):
+    def test_empty_list_can_be_tensorized(self):
+        token_indexers = {"tokens": SingleIdTokenIndexer()}
+        tokenizer = WordTokenizer()
+        tokens = tokenizer.tokenize("Foo")
+        text_field = TextField(tokens, token_indexers)
+        list_field = ListField([text_field.empty_field()])
+        fields = {'list': list_field}
+        instance = Instance(fields)
+        instance.as_tensor_dict()
+
+    def test_end_to_end(self):
+        reader = SalienceReader()
+        dataset = reader.read(self.FIXTURES_ROOT / 'end_to_end' / 'sample.json')
+        vocab = Vocabulary.from_instances(dataset)
+        model = SalienceModel(vocab)
+        model.eval()
+        iterator = BasicIterator(batch_size=2)
+        iterator.index_with(vocab)
+        batch = next(iterator(dataset, shuffle=False))
+        results = model.forward(**batch)["score"]
+        # For the sample data:
+        # {"body": "This is a test.", "entity_name": "exam", "entity_mentions": ["test", "quiz"]}
+        # {"body": "The dog went on a walk.", "entity_name": "animal", "entity_mentions": ["hound", "puppy"]}
```

Comment:
Sorry, I didn't mean to make you do so much work for this simple thing!  I think you're saying that this will fail (before this PR) if the entity mentions are empty for one instance?  Is that right?

Reply:
Hey @matt-gardner, sorry for the delay on this. I've picked it back up and made some fixes. These tests behave as follows:

`test_empty_list_can_be_tensorized`: Broken on master with torch 0.4. Fixed in this PR.
`test_end_to_end_broken_without_fix`: Broken on master even with torch 1.0. Fixed in this PR.
`test_end_to_end_works_in_master`: Works in master. This is because the mentions for at least one instance in the batch is non-empty.

The trace for the test that would still be broken with torch 1.0 is:

```
allennlp/tests/end_to_end/empty_list_test.py:144: in forward
    embedded_mentions_tokens = self.embedder(entity_mentions)
../../../anaconda3/envs/allennlp/lib/python3.6/site-packages/torch/nn/modules/module.py:489: in __call__
    result = self.forward(*input, **kwargs)
allennlp/modules/text_field_embedders/basic_text_field_embedder.py:123: in forward
    token_vectors = embedder(*tensors)
../../../anaconda3/envs/allennlp/lib/python3.6/site-packages/torch/nn/modules/module.py:489: in __call__
    result = self.forward(*input, **kwargs)
allennlp/modules/token_embedders/embedding.py:133: in forward
    inputs = util.combine_initial_dims(inputs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

tensor = tensor([], size=(1, 1, 0), dtype=torch.int64)

    def combine_initial_dims(tensor: torch.Tensor) -> torch.Tensor:
        """
        Given a (possibly higher order) tensor of ids with shape
        (d1, ..., dn, sequence_length)
        Return a view that's (d1 * ... * dn, sequence_length).
        If original tensor is 1-d or 2-d, return it as is.
        """
        if tensor.dim() <= 2:
            return tensor
        else:
>           return tensor.view(-1, tensor.size(-1))
E           RuntimeError: cannot reshape tensor of 0 elements into shape [-1, 0]
```

Reply:
Ok, thanks for this.  To rephrase your use case: you have sparse inputs that are used as additional features for some prediction, and they are sparse enough that they can be empty for some cases.  It would be silly to try to handle these as `None` in your model; it makes a whole lot more sense to just have a minimally-sized tensor that gets entirely masked and has no effect on the rest of the model.

I was thinking about possible alternatives for this; you could, e.g., try a `MultiLabelField`, but that's probably not a great idea, and definitely wouldn't work if your sparse inputs are sentences.  I can't think of a better solution than what you have.  Sorry for all of the hassle, I just didn't understand the use case well enough before seeing the code.

Reply:
Another alternative would be to try to explicitly label something as a `Sparse` field, or something, but that's probably more work than it's worth.  Easier to just let the `List` field handle this.

Reply:
Yep, your rephrasing is spot on. :) And no worries about the hassle! Handling/using padding properly strikes me as an area where care is especially warranted.